llvm/lib/ObjectYAML/GOFFEmitter.cpp

```diff
@@ -0,0 +1,453 @@
+      return;
+    }
+  } else
+    Size = Sec.DataLength;
+
+  GW.newRecord(GOFF::RT_TXT, GOFF::PayloadLength - TXTMaxDataLength + Size);
+  GW << binaryBe(Sec.TextStyle)                // Text Record Style
+     << binaryBe(Sec.SymbolID)                 // Element ESDID
+     << binaryBe(uint32_t(0))                  // Reserved
+     << binaryBe(Sec.Offset)                   // Offset
+     << binaryBe(Sec.TrueLength)               // Text Field True Length
+     << binaryBe(Sec.TextEncoding)             // Text Encoding
+     << binaryBe(static_cast<uint16_t>(Size)); // Data Length
+  if (Sec.Data)
+    GW << *Sec.Data; // Data
+  else
+    GW << zeros(Size);
+}
+
+void GOFFState::writeRelocationDirectory(GOFFYAML::Relocations Rels) {
+  size_t Size = 0;
+  for (const llvm::GOFFYAML::Relocation &Rel : Rels.Relocs) {
+    Size += 8;
+    if (!(Rel.Flags & GOFF::RLD_Same_RID))
+      Size += 4;
+    if (!(Rel.Flags & GOFF::RLD_Same_PID))
+      Size += 4;
+    if (!(Rel.Flags & GOFF::RLD_Same_Offset)) {
+      if (Rel.Flags & GOFF::RLD_Offset_Length)
+        Size += 8;
+      else
+        Size += 4;
+    }
+    if (Rel.Flags & GOFF::RLD_EA_Present)
+      Size += 8;
+  }
+  if (Size > 0xffffULL) {
+    reportError("Relocation directory is too large: " + Twine(Size));
+    return;
+  }
+
+  GW.newRecord(GOFF::RT_RLD, Size + 3);
+  GW << binaryBe(uint8_t(0))      // Reserved
+     << binaryBe(uint16_t(Size)); // Length of relocation data
+  for (const llvm::GOFFYAML::Relocation &Rel : Rels.Relocs) {
+    GW << binaryBe(uint8_t(Rel.Flags)) // Flags, byte 1
+       << binaryBe(uint8_t(Rel.ReferenceType << 4 | Rel.ReferentType)) //
+       << binaryBe(uint8_t(Rel.Action << 1 | (Rel.Flags >> 8)))        //
+       << zeros(1)                                              // Reserved
+       << binaryBe(Rel.TargetFieldByteLength)                   //
+       << binaryBe(uint8_t(Rel.BitLength << 4 | Rel.BitOffset)) //
+       << zeros(2);                                             // Reserved
+    if (!(Rel.Flags & GOFF::RLD_Same_RID))
+      GW << binaryBe(Rel.RPointer);
+    if (!(Rel.Flags & GOFF::RLD_Same_PID))
+      GW << binaryBe(Rel.PPointer);
+    if (!(Rel.Flags & GOFF::RLD_Same_Offset)) {
+      if (Rel.Flags & GOFF::RLD_Offset_Length)
+        GW << binaryBe(Rel.Offset);
+      else
+        GW << binaryBe(static_cast<uint32_t>(Rel.Offset));
+    }
+    if (Rel.Flags & GOFF::RLD_EA_Present) {
+      GW << binaryBe(Rel.ExtAttrID);
+      GW << binaryBe(Rel.ExtAttrOffset);
+    }
+  }
+}
+
+void GOFFState::writeEnd() {
+  GW.newRecord(GOFF::RT_END, GOFF::PayloadLength);
+  GW << binaryBe(uint8_t(0)) // No entry point
+     << binaryBe(uint8_t(0)) // No AMODE
+     << zeros(3)             // Reserved
+     << binaryBe(GW.logicalRecords());
+  // No entry point yet. Automatically fill remaining space with zero bytes.
+  GW.finalize();
+}
+
+bool GOFFState::writeObject() {
+  // We follow a strict, recommended order:
```

Comment:
It's true that GOFF files _should_ follow that order, but this tool doesn't actually obey that rule (which is good, because it lets us test error states) so imo we can actually delete this comment.